### PR TITLE
Fix averaging of LCD touchscreen presses to exclude invalid readings

### PR DESCRIPTION
**Pull request by dvosully**
The drivers for the NS2016 SPI Resistive touchscreen driver and the XPT2046 SPI Resistive touchscreen driver apply a 4 sample average to the touch readings. This means that if a touch is released during the sampling interval, the returned position is corrupted by invalid readings. This means that often when releasing the touchscreen, the position will get 'dragged' or 'torn' towards one corner, and thus presses often trigger a different object than intended.

This patch changes the behaviour to validate the samples individually, rejecting invalid samples, and if there are more than 50% invalid samples they will return no touch. This stabilises the touch behaviour to ensure that the pressed object is the one intended.

### CODE REVIEW

File: components/display/touch_panel/ns2016/ns2016.c [hunk 1]
```diff
@@ -159,19 +159,31 @@ esp_err_t ns2016_get_rawdata(uint16_t *x, uint16_t *y)
...
 
-    aveX /= NS2016_SMP_SIZE;
-    aveY /= NS2016_SMP_SIZE;
+    // If we don't have at least 50% valid samples, there was no valid touch.
+    if (valid_count >= (NS2016_SMP_SIZE / 2)) {
+        aveX /= valid_count;
+        aveY /= valid_count;
+    } else {
+        aveX = 1;
+        aveY = 1;
```
Comment (Lzw655) [inline components/display/touch_panel/ns2016/ns2016.c:185]:
```suggestion
        aveX = TOUCH_SAMPLE_INVALID;
        aveY = TOUCH_SAMPLE_INVALID;
```

File: components/display/touch_panel/xpt2046/xpt2046.c [hunk 2]
```diff
@@ -157,19 +157,31 @@ esp_err_t xpt2046_get_rawdata(uint16_t *x, uint16_t *y)
...
 
-    aveX /= XPT2046_SMP_SIZE;
-    aveY /= XPT2046_SMP_SIZE;
+    // If we don't have at least 50% valid samples, there was no valid touch.
+    if (valid_count >= (XPT2046_SMP_SIZE / 2)) {
+        aveX /= valid_count;
+        aveY /= valid_count;
+    } else {
+        aveX = 1;
+        aveY = 1;
```
Comment (Lzw655) [inline components/display/touch_panel/xpt2046/xpt2046.c:183]:
```suggestion
        aveX = TOUCH_SAMPLE_INVALID;
        aveY = TOUCH_SAMPLE_INVALID;
```

**Review by Lzw655 — COMMENTED** [review]
@dvosully Hi, it's useful to exclude invalid points before averaging. Thanks for your contribution.

I think maybe it's better to use a macro like `TOUCH_SAMPLE_INVALID` instead of a single `1`.

**Comment by dvosully** [comment]
I agree it is much better to use a macro. A zero value suffices, I am not sure why I used a value of 1, what matters is that the value is outside the range TOUCH_SAMPLE_MIN to TOUCH_SAMPLE_MAX. I have changed it to a macro with a value of 0.

**Review by Lzw655 — APPROVED** [review]
It is all good and can be merged.

**Pull request closed without merging**